File: devserver.js
const { createServer } = require('http');
const { readFile, existsSync, lstatSync, readFileSync } = require('fs');
const { extname, join } = require('path');


const server = createServer((request, response) => {
    if (request.url === '/') {
        response.writeHead(200, { 'Content-Type': 'text/html' }).end(getIndex(), 'utf-8')
    }

    const filePath = `./${request.url}`
    const contentType = getContentType(extname(filePath));

    if (existsSync(filePath) && lstatSync(filePath).isFile()) {
        const content = readFileSync(filePath, 'utf-8');
        response.writeHead(200, { 'Content-Type': contentType }).end(content, 'utf-8');
    } else {
        response.writeHead(404, { 'Content-Type': contentType }).end("File not found", 'utf-8');
    }
})

server.listen(3000, () => {
    console.log('listening on localhost:3000')
});


function getIndex() {
    return `<html>
    <head>
        <title>Devserver</title>
    </head>
    <body>
        <div id="app"></div>
        <script type="module" src="./main.js"></script>
    </body>
</html>
    `
}

function getContentType(/** @type {string} */extname) {
    switch (extname) {
        case '.js': return 'text/javascript';
        case '.css': return 'text/css';
        case '.json': return 'application/json';
        case '.png':
        case '.jpg': return 'image/jpg';
        default: return 'text/html';
    }
}

File: devserver.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
)

var index = `<html>
<head>
	<title>Devserver</title>
</head>
<body>
	<div id="app"></div>
	<script type="module" src="./main.js"></script>
</body>
</html>
`

func main() {
	// root := flag.String("root", "", "The root path that should be served (Required)")
	port := flag.Int("port", 3000, "The port where it should be served")
	flag.Parse()

	mux := http.NewServeMux()
	mux.HandleFunc("/", func(res http.ResponseWriter, req *http.Request) {
		if req.URL.Path == "/" {
			fmt.Fprint(res, index)
			return
		}
		http.ServeFile(res, req, req.URL.Path[1:])
	})

	server := http.Server{
		Handler: mux,
		Addr:    fmt.Sprintf(":%v", *port),
	}

	fmt.Printf("\n🌍 Starting server at %s\n", fmt.Sprintf("http://localhost:%d/", *port))
	log.Fatal(server.ListenAndServe())
}
